### ISSUE TITLE
Provide own SSL_CTX

### ISSUE THREAD

**Issue by nfrmtkr** [civetweb/civetweb#526]
We use our own open SSL configuration and initialization infrastructure in order to have a common way of configuration and behaviour for various SSL sockets (not only civetweb). In past we patched civetweb set_ssl_option() function by setting ctx->ssl_ctx = ctx->user_data. The user_data we've set to our initialized SSL_CTX pointer. Of cource the uninitialize_ssl() function requires patches as well in order to avoid deleting the SSL_CTX.

I've tried to use the init_ssl callback, but this is used for certificate creation, only. With this I cannot replace the SSL_CTX. What is the best way of adding selfmaintained SSL_CTX to civetweb. 

Rough idea:  
Add a new callback 

int (*init_ssl_ctx)( void **ssl_context, void *user_data );

Maybe user_data isn't required. If this callback returns 0 then civetweb creates the context. If this callback returns 1 then the ssl_context pointer is set by the callback function. The return code -1 inidcats an error.

This callback will be called in set_ssl_option right after the check for if (!ctx).
An int ssl_ctx_extern member will be added mg_context structure. It inidcates if the SSL_CTX has been created outside civetweb. This will be checked in uninitialize_ssl() in order to avoid deleteing the ssl_ctx stored in mg_context structure.

If we aggree about the way support can be added, I would prepare a pull request. 
Please let me know about your ideas and suggestions.

**Comment by bel2125** [comment]
I don't know if we really need an additional callback - this callback does not really cooperate with the existing `init_ssl` context. In a program you are either using your context or the context created CivetWeb, and you know that during compile time, so you don't need to switch at runtime, right?

So probably the easiest way to do this is a `#define SSL_CTX_EXTERNAL` (or something like this), and if this define is set we just take this value and skip everything in `set_ssl_option` starting from `SSL_CTX_new` (the `initialize_ssl` is already skipped, if `SSL_ALREADY_INITIALIZED` is set).

**Comment by nfrmtkr** [comment]
Yes, I don't need to switch it during runtime. I create SSL_CTX before mg_start and I destroy it after mg_stop. But how we pass this external SSL_CTX to civetweb? The #define will be used in uninitialize_ssl() , too.

**Comment by bel2125** [comment]
Do you have just one SSL_CTX per process, or do you need to have two or more different SSL_CTX for two different server instances running in the same process?

**Comment by nfrmtkr** [comment]
We have more than one SSL_CTX in one process. We open a few SSL enabled server sockets for proprietary protocols with the ACE (adaptive communication environment) library. For this we set up thread specific SSL_CTX for each service.

**Comment by bel2125** [comment]
Then, the callback you suggested seems to be the best solution indeed.

You should add a function to `mg_callbacks`, including the user_data argument:
```C
int (*get_external_ssl_ctx)( void **ssl_ctx, void *user_data );
```

- A return value of `0` would mean civetweb should continue to create the context, just as if the callback would not be present. The value in *ssl_ctx when the function returns is ignored.
- A return value of `-1` would mean, abort the start. The value in *ssl_ctx when the function returns is ignored.
- For a return value `< -1`, the behaviour is undefined, would be initially implemented like -1, but might change in future.
- A return value or `1` would mean, copy the value from *ssl_ctx to the civetweb context, instead of calling `SSL_CTX_new` in `set_ssl_option`
... but then there are two options: 
  - either continue with `set_ssl_option` and perform the `SSL_CTX_set_options` 
  - or return 1 (success) immediately and return from `set_ssl_option`

Probably you want to return immediately, so civetweb does not change any of your options. Some CivetWeb option parameters (like ciphers, protocol version, ...) are ignored in this case. But then it's the responsibility of the callback to set options compatible with civetweb, and it might be required for to check this compatibility of options for each civetweb version.

**Comment by nfrmtkr** [comment]
I've implemeted the new callback and it works for my use case. I don't execute any of the set_ssl_option() calls. If one needs this in future, it may be added by returning 2 in the callback indicating that the context is provided and civetweb's set_ssl_option calls are executed.

How do your want my changes deflivered. A pull reqiuest or a copy of my civetweb.c and civetweb.h files?

**Comment by bel2125** [comment]
A pull request would be fine.
Thanks.

**Comment by bel2125** [comment]
This is solved with PR #533
